File: Leetcode/322. Coin Change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp=[10000000]*(amount+1)
        if amount==0:
            return 0
        for coin in coins:
            if coin<=amount:
                dp[coin]=1
        for coin in coins:
            for i in range(amount+1):
                if coin<=i:
                    dp[i]=min(dp[i],1+dp[i-coin])
        if dp[amount]==10000000:
            return -1
        return dp[amount]
        

File: Leetcode/322. Coin Change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1,10000000);
        if(amount==0){return 0;}
        int n=coins.size();
        for(int i=0;i<n;i++){
            if(coins[i]<=amount){
                dp[coins[i]]=1;
            }
        }
        for(int coin:coins){
            for(int i=1;i<=amount;i++){
                if(i>=coin){
                    dp[i]=min(dp[i],1+dp[i-coin]);
                }
            }
        }
        return dp[amount]==10000000?-1:dp[amount];
    }
};
